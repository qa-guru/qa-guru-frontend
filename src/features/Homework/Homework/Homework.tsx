import React from "react";
import { IHomeworkAnswer } from "./Homework.types";
import { Box, Stack, Typography } from "@mui/material";
import Avatar from "@mui/material/Avatar";
import { ReactComponent as OvenClock } from "../../../icons/ovenclock.svg";
import { ReactComponent as Search } from "../../../icons/search.svg";
import { ReactComponent as Done } from "../../../icons/done.svg";
import { format, parseISO } from "date-fns";
import SerializeAnswer from "./SerializeAnswer";

const style = {
  paper: { padding: "20px", mt: "40px" },
  wrapper: {
    width: "max-content",
    mt: "20px",
  },
  avatar: {
    width: 40,
    height: 40,
  },
};

const Homework: React.FC<IHomeworkAnswer> = (props) => {
  const { data } = props;

  let icon;
  let date;
  let status;
  let createDate;
  let createTime;

  if (data?.homeWorkByStudentAndLecture?.creationDate) {
    createDate = format(
      parseISO(data?.homeWorkByStudentAndLecture?.creationDate),
      "dd.MM.yyyy"
    );

    createTime = format(
      parseISO(data?.homeWorkByStudentAndLecture?.creationDate),
      "HH:mm"
    );
  }

  switch (data?.homeWorkByStudentAndLecture?.status) {
    case "NEW":
      icon = <OvenClock />;
      date = null;
      status = "Ожидает проверки";
      break;
    case "IN_REVIEW":
      icon = <Search />;
      date = format(
        parseISO(data?.homeWorkByStudentAndLecture?.startCheckingDate),
        "dd.MM.yyyy"
      );
      status = "На проверке";
      break;
    case "APPROVED":
      icon = <Done />;
      date = format(
        parseISO(data?.homeWorkByStudentAndLecture?.endCheckingDate),
        "dd.MM.yyyy"
      );
      status = "Принято";
  }

  return (
    <>
      <Stack spacing={3} direction="row" alignItems="center">
        <Typography variant="h5">Ответ на задание</Typography>
        <Stack direction="row" spacing={1.3} alignItems="center">
          {icon}
          <Box>
            <Typography variant="subtitle1">{status}</Typography>
            <Typography variant="subtitle2">{date}</Typography>
          </Box>
        </Stack>
      </Stack>

      <Stack
        spacing={1.7}
        direction="row"
        alignItems="center"
        sx={style.wrapper}
      >
        <Avatar
          sx={style.avatar}
          alt="Remy Sharp"
          src="/static/images/avatar/1.jpg"
        />
        <Box>
          <Typography variant="subtitle1">
            {data?.homeWorkByStudentAndLecture?.student?.firstName}{" "}
            {data?.homeWorkByStudentAndLecture?.student?.lastName}
          </Typography>
          <Stack direction="row" spacing={3}>
            <Typography variant="subtitle2">{createDate}</Typography>
            <Typography variant="subtitle2">{createTime}</Typography>
          </Stack>
        </Box>
      </Stack>

      <SerializeAnswer answer={data?.homeWorkByStudentAndLecture?.answer!} />
    </>
  );
};

export default Homework;
