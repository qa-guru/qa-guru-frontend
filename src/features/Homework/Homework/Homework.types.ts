import { HomeWorkByStudentAndLectureQuery } from "../../../generated/graphql";

export interface IHomeworkAnswer {
  data: HomeWorkByStudentAndLectureQuery;
}
