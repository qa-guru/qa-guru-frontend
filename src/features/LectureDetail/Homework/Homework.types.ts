import { HomeWorkByLectureQuery } from "../../../api/graphql/generated/graphql";

export interface IHomework {
  dataHomeWorkByLecture: HomeWorkByLectureQuery;
}
