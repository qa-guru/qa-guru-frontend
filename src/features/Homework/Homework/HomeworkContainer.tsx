import React from "react";
import Homework from "./Homework";
import { useHomeWorkByStudentAndLectureQuery } from "../../../api/graphql/homework/homeWorkByStudentAndLecture";
import Spinner from "../../../shared/Spinner";
import { useParams } from "react-router-dom";
import { useUserIdQuery } from "../../../api/graphql/user/userId";

const HomeworkContainer: React.FC = () => {
  const { lectureId } = useParams();
  const { data: dataUserId } = useUserIdQuery();
  const { data, loading } = useHomeWorkByStudentAndLectureQuery({
    variables: { studentId: dataUserId?.user?.id!, lectureId: lectureId! },
  });

  if (loading && !data) return <Spinner />;

  return <Homework data={data!} />;
};

export default HomeworkContainer;
