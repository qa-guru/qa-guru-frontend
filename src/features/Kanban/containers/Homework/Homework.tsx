import React from "react";
import {
  useHomeWorkByLectureQuery,
  UserQuery,
} from "../../../../api/graphql/generated/graphql";
import Spinner from "../../../../shared/components/Spinner";
import NoDataErrorMessage from "../../../../shared/components/NoDataErrorMessage";
import Homework from "../../../../shared/Homework";

interface IHomeworkContainer {
  lectureId: string;
  dataUserId: UserQuery;
}

const HomeworkContainer: React.FC<IHomeworkContainer> = ({
  lectureId,
  dataUserId,
}) => {
  const { data: dataHomeWorkByLecture, loading: loadingHomeWorkByLecture } =
    useHomeWorkByLectureQuery({
      variables: { lectureId: lectureId! },
      fetchPolicy: "cache-first",
    });

  if (loadingHomeWorkByLecture) return <Spinner />;
  if (!dataHomeWorkByLecture) return <NoDataErrorMessage />;

  return (
    <Homework
      dataHomeWorkByLecture={dataHomeWorkByLecture}
      dataUserId={dataUserId}
    />
  );
};

export default HomeworkContainer;
