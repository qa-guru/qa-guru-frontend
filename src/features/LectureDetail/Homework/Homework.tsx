import React, { useState } from "react";
import { Box, Button, Paper, Stack, Typography } from "@mui/material";
import Avatar from "@mui/material/Avatar";
import { format, parseISO } from "date-fns";
import HighlightOffIcon from "@mui/icons-material/HighlightOff";
import { styled } from "@mui/material/styles";
import { IHomework } from "./Homework.types";
import UpdateHomework from "./UpdateHomework";
import SendHomeWork from "./SendHomework";
import Comment from "./Comment";
import { ReactComponent as Clock } from "../../../assets/icons/clock.svg";
import { ReactComponent as Search } from "../../../assets/icons/search.svg";
import { ReactComponent as Done } from "../../../assets/icons/done.svg";
import TextSerialization from "../../../shared/TextSerialization";
import { red } from "../../../theme/colors";

const RedHighlightOffIcon = styled(HighlightOffIcon)({
  color: red.main,
});

const style = {
  paper: { padding: { xs: "15px", md: "20px" }, mt: "40px" },
  buttonUpdate: { textTransform: "none", minWidth: "147px", mt: "15px" },
  avatar: {
    width: 40,
    height: 40,
  },
  wrapper: {
    width: "max-content",
    mt: "20px",
  },
};

const Homework: React.FC<IHomework> = ({ dataHomeWorkByLecture }) => {
  const {
    status,
    startCheckingDate,
    endCheckingDate,
    mentor,
    student,
    creationDate,
    answer,
    id,
  } = dataHomeWorkByLecture.homeWorkByLecture! || {};

  const [openHomeWorkEdit, setOpenHomeWorkEdit] = useState<boolean>(false);
  const DATE_FORMAT = "dd.MM.yyyy";

  let icon;
  let date;
  let statusText;
  let homework;

  switch (status) {
    case "NEW":
      icon = <Clock />;
      date = null;
      statusText = "Ожидает проверки";
      break;
    case "IN_REVIEW":
      icon = <Search />;
      date = format(parseISO(startCheckingDate), DATE_FORMAT);
      statusText = "На проверке";
      break;
    case "APPROVED":
      icon = <Done />;
      date = format(parseISO(endCheckingDate), DATE_FORMAT);
      statusText = "Принято";
      break;
    case "NOT_APPROVED":
      icon = <RedHighlightOffIcon />;
      date = format(parseISO(endCheckingDate), DATE_FORMAT);
      statusText = "Не принято";
      break;
  }

  if (status && !openHomeWorkEdit) {
    homework = <TextSerialization text={answer!} />;
  } else if (status && openHomeWorkEdit) {
    homework = (
      <UpdateHomework
        answer={answer}
        setOpenHomeWorkEdit={setOpenHomeWorkEdit}
        id={id}
      />
    );
  } else {
    homework = <SendHomeWork />;
  }

  return (
    <Paper sx={style.paper}>
      <Stack spacing={3} direction="row" alignItems="center">
        <Typography variant="h5">Ответ на задание</Typography>
        <Stack direction="row" spacing={1} alignItems="center">
          {icon}
          <Box>
            <Typography variant="subtitle1">{statusText}</Typography>
            <Typography variant="subtitle2">{date}</Typography>
          </Box>
          {status === ("IN_REVIEW" || "APPROVED" || "NOT_APPROVED") && (
            <Stack direction="row" alignItems="center" spacing={1}>
              <Avatar
                sx={style.avatar}
                alt="Remy Sharp"
                src="/static/images/avatar/1.jpg"
              />
              <Typography variant="subtitle1">
                {mentor?.firstName} {mentor?.lastName}
              </Typography>
            </Stack>
          )}
        </Stack>
      </Stack>

      {status && (
        <Stack
          spacing={1.7}
          direction="row"
          alignItems="center"
          sx={style.wrapper}
        >
          <Avatar
            sx={style.avatar}
            alt="Remy Sharp"
            src="/static/images/avatar/1.jpg"
          />
          <Box>
            <Typography variant="subtitle1">
              {student?.firstName} {student?.lastName}
            </Typography>
            <Stack direction="row" justifyContent="space-between">
              <Typography variant="subtitle2">
                {format(parseISO(creationDate!), "dd.MM.yyyy '|' HH:mm")}
              </Typography>
            </Stack>
          </Box>
        </Stack>
      )}

      <Box mt="7px">{homework}</Box>
      {!openHomeWorkEdit && status && (
        <Button
          onClick={() => setOpenHomeWorkEdit(true)}
          sx={style.buttonUpdate}
          variant="contained"
        >
          Редактировать
        </Button>
      )}

      {status && <Comment id={id!} />}
    </Paper>
  );
};

export default Homework;
