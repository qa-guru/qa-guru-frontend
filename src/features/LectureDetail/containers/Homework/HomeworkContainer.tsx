import React, { useContext } from "react";
import Homework from "../../../../shared/Homework";
import {
  useHomeWorkByLectureQuery,
  useUserIdQuery,
} from "../../../../api/graphql/generated/graphql";
import { LectureIdContext } from "../../context/LectureIdContext";
import NoDataErrorMessage from "../../../../shared/components/NoDataErrorMessage";
import Spinner from "../../../../shared/components/Spinner";

const HomeworkContainer: React.FC = () => {
  const lectureId = useContext(LectureIdContext);
  const { data: dataUserId, loading: loadingUserId } = useUserIdQuery();

  const { data: dataHomeWorkByLecture, loading: loadingHomeWorkByLecture } =
    useHomeWorkByLectureQuery({
      variables: { lectureId: lectureId! },
      fetchPolicy: "cache-first",
    });

  if (loadingUserId || loadingHomeWorkByLecture) return <Spinner />;
  if (!dataHomeWorkByLecture || !dataUserId) return <NoDataErrorMessage />;

  return (
    <Homework
      dataHomeWorkByLecture={dataHomeWorkByLecture.homeWorkByLecture!}
      dataUserId={dataUserId!}
    />
  );
};

export default HomeworkContainer;
