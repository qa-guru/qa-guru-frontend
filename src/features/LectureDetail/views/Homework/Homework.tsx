import React, { useContext } from "react";
import { useModal } from "react-modal-hook";
import { Button, Dialog, DialogContent, Paper, Box } from "@mui/material";
import ClearIcon from "@mui/icons-material/Clear";
import { IHomework } from "./Homework.types";
import { style } from "./styles";
import Comments from "../../containers/Comments";
import CommentsLimited from "../CommentsLimited";
import CommentsPagination from "../CommentsPagination";
import { LectureIdContext } from "../../context/LectureIdContext";
import HomeworkItem from "../HomeworkItem";

const Homework: React.FC<IHomework> = (props) => {
  const { dataHomeWorkByLecture, dataUserId } = props;
  const lectureId = useContext(LectureIdContext);

  const [showModal, hideModal] = useModal(({ in: open }) => (
    <Dialog open={open} onClose={hideModal} maxWidth="xl" fullWidth>
      <DialogContent sx={style.scrollContainer} id="scroll-container">
        <ClearIcon sx={style.clearIcon} onClick={hideModal} />
        <LectureIdContext.Provider value={lectureId}>
          <HomeworkItem
            dataHomeWorkByLecture={dataHomeWorkByLecture}
            dataUserId={dataUserId}
          />
          <Comments id={dataHomeWorkByLecture?.id!}>
            <CommentsPagination />
          </Comments>
        </LectureIdContext.Provider>
      </DialogContent>
    </Dialog>
  ));

  return (
    <Paper sx={style.paper}>
      <HomeworkItem
        dataHomeWorkByLecture={dataHomeWorkByLecture}
        dataUserId={dataUserId}
      />
      {dataHomeWorkByLecture?.id! && (
        <>
          <Comments id={dataHomeWorkByLecture?.id!}>
            <CommentsLimited />
          </Comments>
          <Box mt={2}>
            <Button sx={style.button} variant="contained" onClick={showModal}>
              Показать другие комментарии
            </Button>
          </Box>
        </>
      )}
    </Paper>
  );
};

export default Homework;
