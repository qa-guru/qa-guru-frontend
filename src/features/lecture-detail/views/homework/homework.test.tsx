import { MemoryRouter } from "react-router-dom";
import { render } from "test/utils-test";
import { userId } from "shared/mocks/user-id.mock";
import { studentHomeWorkDto } from "shared/mocks/student-homework-dto.mock";
import Homework from "./homework";

describe("Homework", () => {
  it("renders correctly", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Homework
          dataHomeWorkByLecture={studentHomeWorkDto}
          dataUserId={userId}
        />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
