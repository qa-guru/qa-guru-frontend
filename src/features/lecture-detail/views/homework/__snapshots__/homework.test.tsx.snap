// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Homework > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-ay1hcn-MuiPaper-root"
  >
    <div
      class="MuiStack-root css-d3d1h5-MuiStack-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-xlem1t-MuiTypography-root"
      >
        Ответ на задание
      </h5>
      <div
        class="MuiStack-root css-19iseno-MuiStack-root"
      >
        <svg
          fill="none"
          height="20"
          viewBox="0 0 20 20"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.99 0C4.47 0 0 4.48 0 10C0 15.52 4.47 20 9.99 20C15.52 20 20 15.52 20 10C20 4.48 15.52 0 9.99 0ZM10 18C5.58 18 2 14.42 2 10C2 5.58 5.58 2 10 2C14.42 2 18 5.58 18 10C18 14.42 14.42 18 10 18Z"
            fill="black"
            fill-opacity="0.56"
          />
          <path
            d="M10.5 5H9V11L14.25 14.15L15 12.92L10.5 10.25V5Z"
            fill="black"
            fill-opacity="0.56"
          />
        </svg>
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 css-3c3maz-MuiTypography-root"
          >
            Ожидает проверки
          </p>
        </div>
      </div>
      <div
        class="MuiStack-root css-txf2s8-MuiStack-root"
      />
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiBox-root css-1kawqjx"
    >
      <div
        class="MuiStack-root css-19iseno-MuiStack-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1d0j7ho-MuiAvatar-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 css-4ozz3r-MuiTypography-root"
          >
            fl
          </h6>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1 css-v3tflm-MuiTypography-root"
          >
            firstName lastName
          </h6>
          <div
            class="MuiStack-root css-1je0tgk-MuiStack-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 css-4ozz3r-MuiTypography-root"
            >
              10.10.2021 | 12:11
            </h6>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1kawqjx"
    >
      <div
        class="MuiBox-root css-i6bazn"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle1 css-tm4gf8-MuiTypography-root"
        >
          answer
          <br />
        </h6>
      </div>
    </div>
    <div
      class="MuiStack-root css-ba457o-MuiStack-root"
    >
      <div
        class="MuiStack-root css-vx72w3-MuiStack-root"
      >
        <div
          class="MuiStack-root css-16jryck-MuiStack-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-a4z31c-MuiSkeleton-root"
          />
          <div
            class="MuiStack-root css-1gcuwd7-MuiStack-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-xs38fd-MuiSkeleton-root"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-xs38fd-MuiSkeleton-root"
            />
          </div>
        </div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rounded MuiSkeleton-pulse css-8hw9pq-MuiSkeleton-root"
        />
      </div>
      <div
        class="MuiStack-root css-vx72w3-MuiStack-root"
      >
        <div
          class="MuiStack-root css-16jryck-MuiStack-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse css-a4z31c-MuiSkeleton-root"
          />
          <div
            class="MuiStack-root css-1gcuwd7-MuiStack-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-xs38fd-MuiSkeleton-root"
            />
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-xs38fd-MuiSkeleton-root"
            />
          </div>
        </div>
        <span
          class="MuiSkeleton-root MuiSkeleton-rounded MuiSkeleton-pulse css-8hw9pq-MuiSkeleton-root"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1yyn4lv-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Показать другие комментарии
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
